### PR TITLE
Add common canvas API

### PR DESCRIPTION
**Pull request by newsch**
This adds a common API for us to develop off of individually.

We talked IRL about some final steps before merging:
- [x] rename functions to be prefixed with `canvas`
- [x] reorder function parameters to match ncurses standards
- [x] pull out the "test" code in the commented main into a formal test suite

### CODE REVIEW

**Comment by newsch** [comment]
[Add API for canvases](https://trello.com/c/bx4kZoUR/15-add-api-for-canvases)

**Comment by newsch** [comment]
@MatthewBeaudouinLafon @labseven I think this is good, reviews would be appreciated.

~~@MatthewBeaudouinLafon I ended up fixing the Makefile here, that might add some merge/rebase conflicts on your branch.~~

**Comment by newsch** [comment]
I rebased and cleaned up some commits, _now_ it's ready.

**Comment by newsch** [comment]
Rebased off of new master

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+language: c
+compiler: gcc
+
+before_script: cd src
+
+script: make test
```
Comment (labseven) [inline .travis.yml]:
```suggestion
script: make test

```

File: src/canvas.c [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+ * on output error from fprintf
+ */
+int canvas_print(Canvas *canvas)
+{
+    return canvas_fprint(stdout, canvas);
+}
+
+/* Convert a canvas object into a character buffer
+ *
+ * A canvas of size nxm requires a buffer of size n*m bytes (chars).
```
Comment (labseven) [inline src/canvas.c]:
```suggestion
 * A canvas of size n rols and m cols requires a buffer of size n*m bytes (chars).
```
Do x and * mean the same thing in this sentence? nxm took me a bit to parse

File: src/canvas.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+char canvas_gcharyx(Canvas *canvas, int y, int x);
+char canvas_gchari(Canvas *canvas, int i);
+int canvas_load_str(Canvas *canvas, char *str);
+int canvas_fprint(FILE *stream, Canvas *canvas);
+int canvas_print(Canvas *canvas);
+int canvas_serialize(Canvas *canvas, char *buf);
+void canvas_deserialize(char *bytes, Canvas *canvas);
+int canvas_eq(Canvas *a, Canvas *b);
+
+#endif
```
Comment (labseven) [inline src/canvas.h]:
```suggestion
#endif

```

File: src/canvas_test.c [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    canvas_free(c4);
+}
+
+// Tests to make sure the setup/teardown and eq functions will work
+MU_TEST_SUITE(canvas_bootstrapping) {
+    MU_RUN_TEST(test_canvas_new);
+    MU_RUN_TEST(test_canvas_load_str);
+    MU_RUN_TEST(test_canvas_eq);
+}
+
```
Comment (labseven) [inline src/canvas_test.c:76]:
```suggestion

// Begin canvas_main tests
```
This jump confused me a little, I was thinking, "Why suddenly define some functions?"

File: src/canvas_test.c [hunk 5]
```diff
@@ -0,0 +1,158 @@
+#include "minunit.h"
+#include "canvas.h"
+
+static int cols = 2;
+static int rows = 3;
+static char load_str[] = "012345";
+
+static Canvas *c1, *c2;
+
```
Comment (labseven) [inline src/canvas_test.c:9]:
```suggestion

// Begin canvas_bootstrapping tests
```

Below, the jump to canvas_main tests confused me, so I suggested a comment. Adding this one in to be consistent.

**Review by labseven — APPROVED** [review]
Looks great, made some comment suggestions.
Ship it!

**Review by MatthewBeaudouinLafon — APPROVED** [review]

File: src/canvas.c (hunk 2 above)
Comment (newsch) [reply src/canvas.c]: I meant it as "n by m"

**Comment by labseven** [comment]
Ship it!

**Pull request merged**